pennylane/beta/tapes/rev.py

```diff
@@ -0,0 +1,204 @@
+        Unmeasured wires are contracted, and a scalar is returned.
+
+        Args:
+            vec1 (array[complex]): a length :math:`2^N` statevector
+            obs (.Observable): a PennyLane observable
+            vec2 (array[complex]): a length :math:`2^N` statevector
+            device (.QubitDevice): the device used to compute the matrix elements
+        """
+        # pylint: disable=protected-access
+        mat = device._reshape(obs.matrix, [2] * len(obs.wires) * 2)
+        wires = obs.wires
+
+        vec1_indices = ABC[: device.num_wires]
+
+        obs_in_indices = "".join(ABC_ARRAY[wires.tolist()].tolist())
+        obs_out_indices = ABC[device.num_wires : device.num_wires + len(wires)]
+        obs_indices = "".join([obs_in_indices, obs_out_indices])
+
+        vec2_indices = reduce(
+            lambda old_string, idx_pair: old_string.replace(idx_pair[0], idx_pair[1]),
+            zip(obs_in_indices, obs_out_indices),
+            vec1_indices,
+        )
+
+        einsum_str = "{vec1_indices},{obs_indices},{vec2_indices}->".format(
+            vec1_indices=vec1_indices,
+            obs_indices=obs_indices,
+            vec2_indices=vec2_indices,
+        )
+
+        return device._einsum(einsum_str, device._conj(vec1), mat, vec2)
+
+    def jacobian(self, device, params=None, **options):
+        # The parameter_shift_var method needs to evaluate the circuit
+        # at the unshifted parameter values; the pre-rotated statevector is then stored
+        # self._state attribute. Here, we set the value of the attribute to None
+        # before each Jacobian call, so that the statevector is calculated only once.
+        self._state = None
+        return super().jacobian(device, params, **options)
+
+    def analytic_pd(self, idx, device, params, **options):
+        t_idx = list(self.trainable_params)[idx]
+        op = self._par_info[t_idx]["op"]
+        p_idx = self._par_info[t_idx]["p_idx"]
+
+        # The reversible tape only support differentiating
+        # expectation values of observables for now.
+        for m in self.measurements:
+            if (
+                m.return_type is qml.operation.Variance
+                or m.return_type is qml.operation.Probability
+            ):
+                raise ValueError(
+                    f"{m.return_type} is not supported with the reversible gradient method"
+                )
+
+        # The reversible tape only supports the RX, RY, RZ, and Rot operations for now:
+        #
+        # * CRX, CRY, CRZ ops have a non-unitary matrix as generator.
+        #
+        # * PauliRot, MultiRZ, U2, and U3 do not have generators specified.
+        #
+        # TODO: the controlled rotations can be supported by multiplying ``state``
+        # directly by these generators within this function
+        # (or by allowing non-unitary matrix multiplies in the simulator backends)
+
+        if op.name not in ["RX", "RY", "RZ", "Rot"]:

```

Comment:
It was easier to consider the supported gates, rather than the unsupported gates :)